src/maximum_cardinality_matching.jl

```diff
@@ -0,0 +1,118 @@
+"""
+maximum_cardinality_matching(g::Graph)
+
+Given a graph `g` returns a maximum cardinality matching.
+This is the same as running `maximum_weight_matching` with default weights (`1`)
+but is faster without needing a JuMP solver.
+
+Returns MatchingResult containing:
+  - the maximum cardinality that can be achieved
+  - a list of each vertex's match (or -1 for unmatched vertices)
+"""
+function maximum_cardinality_matching(g::AbstractGraph{U}) where U<:Integer
+    n = nv(g)
+    matching = -ones(Int, n)
+    
+    # the number of edges that can possibly be part of a matching
+    max_generally_possible = fld(n, 2)
+
+    # get initial matching
+    matching_len = 0
+    for e in edges(g)
+        # if unmatched
+        if matching[e.src] == -1 && matching[e.dst] == -1
+            matching[e.src] = e.dst
+            matching[e.dst] = e.src
+            matching_len += 1
+        end
+    end
+
+    # if there are at least two free vertices
+    if matching_len < max_generally_possible
+
+        parents = zeros(Int, n)
+        visited = falses(n)
+        
+        @inbounds while matching_len < max_generally_possible
+            cur_level = Vector{Int}()
+            sizehint!(cur_level, n)
+            next_level = Vector{Int}()
+            sizehint!(next_level, n)
```

Comment:
Can we move the allocation of `cur_level` and `next_level` out of the loop, so that we do not have allocations in each iteration?

Reply:
Ah yeah it should be outside. Somehow missed it. Thanks for pointing it out.